samples/ConPTY/MiniTerm/MiniTermCore/Terminal.cs

```diff
@@ -0,0 +1,114 @@
+using Microsoft.Win32.SafeHandles;
+using System.Text;
+using Windows.Win32;
+
+namespace MiniTermCore
+{
+    /// <summary>
+    /// The UI of the terminal. It's just a normal console window, but we're managing the input/output.
+    /// In a "real" project this could be some other UI.
+    /// </summary>
+    internal static class Terminal
+    {
+        private const string CtrlC_Command = "\x3";
+
+        internal enum CtrlTypes : uint
+        {
+            CTRL_C_EVENT = 0,
+            CTRL_BREAK_EVENT,
+            CTRL_CLOSE_EVENT,
+            CTRL_LOGOFF_EVENT = 5,
+            CTRL_SHUTDOWN_EVENT
+        }
+
+        /// <summary>
+        /// Start the pseudoconsole and run the process as shown in 
+        /// https://docs.microsoft.com/en-us/windows/console/creating-a-pseudoconsole-session#creating-the-pseudoconsole
+        /// </summary>
+        /// <param name="command">the command to run, e.g. cmd.exe</param>
+        public static void Run(string command)
+        {
+            using var inputPipe = new PseudoConsolePipe();
+            using var outputPipe = new PseudoConsolePipe();
+            using var pseudoConsole = PseudoConsole.Create(inputPipe.ReadSide, outputPipe.WriteSide, (short)Console.WindowWidth, (short)Console.WindowHeight);
+            using var process = System.Diagnostics.Process.Start(command) ?? throw new Exception();
+
+            // copy all pseudoconsole output to stdout
+            Task.Run(() => CopyPipeToOutput(outputPipe.ReadSide));
+            // prompt for stdin input and send the result to the pseudoconsole
+            Task.Run(() => CopyInputToPipe(inputPipe.WriteSide));
+            // free resources in case the console is ungracefully closed (e.g. by the 'x' in the window titlebar)
+            OnClose(() => DisposeResources(process, pseudoConsole, outputPipe, inputPipe));
+
+            process.WaitForExit();

```

Comment:
Windows 11 26100 includes a new API to make process shutdown safer (`ReleasePseudoConsole`). The API closes all references to the PTY and then you can simply read from the output pipe until the pipe gets closed. That way you don't need to wait for the process exit here.

But given how novel the API is, I'm not sure if it's time for that yet.